File: parent-medic-facility-service/src/main/java/org/example/repository/ParentMedicRepository.java
package org.example.repository;

import lombok.extern.slf4j.Slf4j;
import org.example.entity.ParentMedicFacility;
import org.example.entity.ParentMedicFacilityParam;
import org.example.exception.DataNotFoundException;
import org.example.mapper.ParentMedicMapper;
import org.example.response.HttpResponse;
import org.example.response.RepositoryData;
import org.example.util.QueryUtil;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.r2dbc.core.R2dbcEntityTemplate;
import org.springframework.stereotype.Repository;
import reactor.core.publisher.Mono;

import java.util.List;

@Repository
@Slf4j
public class ParentMedicRepository implements IRepository{

    @Autowired
    private R2dbcEntityTemplate template;

    private ParentMedicMapper mapper = ParentMedicMapper.INSTANCE;

    @Override
    public Mono<RepositoryData<ParentMedicFacility>> save(ParentMedicFacility data) {

        log.info("catch save data: {}", data);

        return this
                .template
                .insert(data)
                .map((d) -> RepositoryData
                        .<ParentMedicFacility>builder()
                        .data(d)
                        .build()
                );
    }

    @Override
    public Mono<RepositoryData<ParentMedicFacility>> get(ParentMedicFacilityParam param) {

        log.info("catch get param: {}", param);

        var query = QueryUtil
                .generateReactiveQuery(param, false);

        return this.template
                .selectOne(query, ParentMedicFacility.class)
                .switchIfEmpty(Mono.error(new DataNotFoundException("parent medic facility data not found")))
                .map((d) -> RepositoryData
                        .<ParentMedicFacility>builder()
                        .data(d)
                        .build()
                );
    }

    @Override
    public Mono<RepositoryData<List<ParentMedicFacility>>> getList(ParentMedicFacilityParam param) {

        var query = QueryUtil
                .generateReactiveQuery(param, false);

        var queryCount = QueryUtil
                .generateReactiveQuery(param, true);

        return Mono.zip(
                this.template
                        .select(query, ParentMedicFacility.class)
                        .collectList()
                        .switchIfEmpty(Mono.just(List.of())),
                this.template
                        .count(queryCount, Long.class)
                        .switchIfEmpty(Mono.just(0L))
        ).map((t2) -> {
            var data = t2.getT1();
            var n = t2.getT2();

            var totalPage = (n / param.getPgParam().getLimit()) + 1;

            return RepositoryData
                    .<List<ParentMedicFacility>>builder()
                    .data(data)
                    .pg(
                            HttpResponse.Pagination
                                    .builder()
                                    .totalPage(totalPage)
                                    .totalElements(n)
                                    .currentPage(param.getPgParam().getOffset().longValue())
                                    .currentElements((long)data.size())
                                    .build()
                    )
                    .build();
        });
    }

    @Override
    public Mono<RepositoryData<ParentMedicFacility>> update(ParentMedicFacilityParam param, ParentMedicFacility data) {
        return this.get(param)
                .map(RepositoryData::getData)
                .flatMap((d) -> {
                    var updateData = this.mapper.update(data, d);
                    return this.save(updateData);
                })
                .map(RepositoryData::getData)
                .map((d) ->
                        RepositoryData
                                .<ParentMedicFacility>builder()
                                .data(d)
                                .build()
                );
    }
}

File: parent-medic-facility-service/src/main/kotlin/repository/ParentMedicRepository.kt
package repository

import entity.ParentMedicFacility
import entity.ParentMedicParam
import mapper.ParentMedicMapper
import org.example.response.HttpResponse.Pagination
import org.example.response.RepositoryData
import org.example.util.QueryUtil
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.data.r2dbc.core.R2dbcEntityTemplate
import reactor.core.publisher.Mono

@org.springframework.stereotype.Repository
class ParentMedicRepository(
    private val r2dbcEntityTemplate: R2dbcEntityTemplate,
    private val log: Logger = LoggerFactory.getLogger(ParentMedicRepository::class.java),
): Repository {

    private var mapper = ParentMedicMapper.INSTANCE

    override fun save(data: ParentMedicFacility): Mono<RepositoryData<ParentMedicFacility>> {

        log.info("catch save data: {}", data)

        return this
            .r2dbcEntityTemplate
            .insert(data)
            .map {
                RepositoryData
                    .builder<ParentMedicFacility>()
                    .data(it)
                    .build()
            }
    }

    override fun get(param: ParentMedicParam): Mono<RepositoryData<ParentMedicFacility>> {

        log.info("catch get param: {}", param)

        val query = QueryUtil
            .generateReactiveQuery(param, false)

        return this.r2dbcEntityTemplate
            .selectOne(query, ParentMedicFacility::class.java)
            .map {
                RepositoryData
                    .builder<ParentMedicFacility>()
                    .data(it)
                    .build()
            }
    }

    override fun getList(param: ParentMedicParam): Mono<RepositoryData<List<ParentMedicFacility>>> {

        log.info("catch get list param: {}", param)

        val query = QueryUtil
            .generateReactiveQuery(param, false)

        val queryCount = QueryUtil
            .generateReactiveQuery(param, true)

        val executeQuery = this.r2dbcEntityTemplate
            .select(query, ParentMedicFacility::class.java)
            .collectList()

        val executeQueryCount = this.r2dbcEntityTemplate
            .count(queryCount, Long::class.java)

        return Mono.zip(
            executeQuery, executeQueryCount
        ).map {
            val list = it.t1
            val n = it.t2

            val totalPage = (n / param.pgParam.limit) + 1

            RepositoryData
                .builder<List<ParentMedicFacility>>()
                .data(list)
                .pg(
                    Pagination
                        .builder()
                        .currentPage(param.pgParam.offset.toLong())
                        .currentElements(list.size.toLong())
                        .totalPage(totalPage)
                        .totalElements(n)
                        .build()
                )
                .build()
        }
    }

    override fun update(param: ParentMedicParam, data: ParentMedicFacility): Mono<RepositoryData<ParentMedicFacility>> {

        log.info("catch update param: {} with data to be updated: {}", param, data)

        val repositoryData = this.get(param)

        return repositoryData
            .map {
                val databaseData = it.data
                this.mapper.updateParentMedic(data, databaseData)
            }
            .flatMap {
                this.save(it)
            }.then(Mono.just(
                RepositoryData
                    .builder<ParentMedicFacility>()
                    .build()
            ))
    }


}